neard/src/runtime.rs

```diff
@@ -892,6 +892,46 @@ impl RuntimeAdapter for NightshadeRuntime {
         Ok(epoch_start_height)
     }
 
+    fn check_sync_hash_on_epoch_boundary(
+        &self,
+        sync_hash: &CryptoHash,
+        last_block_hash: &CryptoHash,
+    ) -> bool {
+        let mut epoch_manager = self.epoch_manager.as_ref().write().expect(POISONED_LOCK_ERR);
+        let block_info = match epoch_manager.get_block_info(sync_hash) {
+            Ok(block_info) => block_info.clone(),
+            Err(_) => return false,
+        };
+
+        // 1. sync_hash must be on the Epoch boundary
+        if block_info.epoch_first_block == *sync_hash {
+            // 2. Epoch of sync_hash can be either the last one or the previous one
+            // 2a. Epoch of sync_hash may be the latest one
+            if sync_hash == last_block_hash {
```

Comment:
Did you mean to compare epochs, not blocks, here?
The comment seems to suggest that you want to compare epochs, and also there's no logic that checks the case when the sync hash is in the current epoch. The next branch covers the case of sync hash being in the previos epoch.